pyat/at/lattice/idtable_element.py

```diff
@@ -159,34 +200,61 @@ def __init__(self,
+        vkickmap = sorted_table(vkickmap_a, rows2sorted_index, 'row')
+        # Reorder kickmap1
+        hkickmap1_a = sorted_table(hkickmap1, cols1sorted_index, 'col')
+        hkickmap1 = sorted_table(hkickmap1_a, rows1sorted_index, 'row')
+        vkickmap1_a = sorted_table(vkickmap1, cols2sorted_index, 'col')
+        vkickmap1 = sorted_table(vkickmap1_a, rows2sorted_index, 'row')
 
         # Field to kick factors
         Brho = 1e9 * Energy/clight
         factor = 1.0/(Brho**2)
         xkick = factor * hkickmap
         ykick = factor * vkickmap
+        # kick1 vars set to zero, not yet implemented
         factor1 = -1.0/(Brho)
-        xtable = table_cols1array
-        ytable = table_rows1array
-
-        # Create element properties
-        kwargs.setdefault('PassMethod', 'IdTablePass')
-        kwargs.setdefault('Filename_in', Filename_in)
-        kwargs.setdefault('Energy',      Energy)
-        kwargs.setdefault('Nslice',      Nslice)
-        kwargs.setdefault('Length',      el_length)
-        kwargs.setdefault('xkick',       xkick)
-        kwargs.setdefault('ykick',       ykick)
-        kwargs.setdefault('xtable',      xtable)
-        kwargs.setdefault('ytable',      ytable)
-
-        # Set Element inherited properties
-        # pyat issue #522
-        # elem = Element('name', PassMethod='IdTablePass', \
-        #                       xkick=x0, ykick=x1, \
-        #                       xtable=x2, ytable=x3, Nslice=x4)
-        super(InsertionDeviceKickMap, self).__init__(family_name, **kwargs)
+        xkick1 = factor1 * hkickmap1
+        ykick1 = factor1 * vkickmap1
+        xtable = table_cols1array.T
+        ytable = table_rows1array.T
+
+        args_dict = {'PassMethod': 'IdTablePass',
+                     'Filename_in': Filename_in,
+                     'Normalization_energy': Energy,
+                     'Nslice': numpy.uint8(Nslice),
+                     'Length': el_length,
+                     'xkick': xkick,
+                     'ykick': ykick,
+                     'xkick1': xkick1,
+                     'ykick1': ykick1,
+                     'xtable': xtable,
+                     'ytable': ytable,
+                     }
+        return args_dict
+
+    def __init__(self, family_name: str, *args, **kwargs):
+        _argnames = ['PassMethod',
+                     'Filename_in',
+                     'Normalization_energy',
+                     'Nslice',
+                     'Length',
+                     'xkick',
+                     'ykick',
+                     'xkick1',
+                     'ykick1',
+                     'xtable',
+                     'ytable']
+        if len(args) < 11:
+            # get data from text file
+            elemargs = self.from_text_file(*args)
+        else:
+            # get data from arguments
+            elemargs = dict(zip(_argnames, args))
+            elemargs.update(kwargs)
```

Comment:
My last comment: you can take the `elemargs.update(kwargs)` statement out of the test, so that it applies in both cases:
```python
    if len(args) < 11:
        # get data from text file
        elemargs = self.from_text_file(*args)
    else:
        # get data from arguments
        elemargs = dict(zip(_argnames, args))
    elemargs.update(kwargs)
    super().__init__(family_name, **elemargs)
```

Thanks for your patience, now it's ok for me !

Reply:
No problem at all. This suggestion is implemented in the last commit.